Technical docs for Snatch developers (#43)
kyma-project/kim-snatch: docs/contributor/README.md

> # Architecture

Reviewer: ```suggestion
# KIM Snatch Architecture
```

> 

Reviewer: How about adding a short overview along with the diagram? Just 1-3 introductory sentences will do.

Author: fixed

|Kubernetes Secret|Created by Gardener Cert-Manager to store the generated CA, TLS key and certificate|
|WebhookConfiguration|Customer Resource of Kubernetes considered by API server to invoke webhooks during request processing|
|API Server|API server of Kubernetes which processes HTTP request and applies resources changes in ETCD|
|Webhook|Webserver intercepting requests before the API server is processing them|
|Webhook HTTPS-Server|Invoked by the API Server and intercepts all incoming requests of the API server.|
|Webhook certs volume|Mount point which makes the data entries of the Secret accessible over the webhook's filesystem|
|Webhook Cert Watcher|Monitoring changes applied on mounted secret. 

## How it works

* Gardener Cert Manager issues a CA, TLS key and certificate. It stores them as entires in a Kubernetes secret.
* A webhook is started (as pod).
   - It mounts the secret to make the CA, TLS key and certificate accessible via the local filesystem.
   - An HTTPS endpoint is exposed, using the mounted TLS certificate for securing incoming connections.

>    - A Cert Watcher triggers an update of the `WebhookConfiguration` and a reload of the HTTPS server whenever an entry in the mounted secret was modified. This keeps the used secrets up-to-date even after they were rotated.|

Reviewer 2: There is a unnecessary `|` character at the end of the line

2. The namespace of a pod has to be labeled with `operator.kyma-project.io/managed-by: kyma`. Only if this is the case, the pod will get a `nodeAffinity` assigned.
3. The assigned `nodeAffinity` is of type `preferredDuringSchedulingIgnoredDuringExecution` . The scheduler follows this affinity with best effort, but not enforcing it. This minimizes the risk that pods won't be scheduled if a suitable node cannot be found. The node affinity is added with a  [`weight`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.32/#preferredschedulingterm-v1-core) of 10 (indicates the priority of this `nodeAffinity`, 0=low, 100=high) which allows the definition of further `nodeAffinity` rules with higher relevance.
4. Already scheduled pods are not considered by the webhook. It interferes only pods which are not yet passed to the Kubernetes scheduler. Means, running pods won't be touched (no `nodeAffinity` will be added to their manifest) and also not restarted.

> ## Example pod manifest with `nodeAffinity`
> This is an example of pod manifest which was intercepted and updated by the mutating webhook:

Reviewer: ```suggestion
See an example of a Pod manifest with `nodeAffinity`, which was intercepted and updated by the mutating webhook:
```

Snatch is part or the Kyma Infrastructure Manager (KIM) and responsible to assign Kyma workloads to the Kyma owned worker pool. This document describes it's architecture and relevant technical flows.

![Snatch Architecture](./assets/architecture-webhook.svg)

1) Gardener cert-manager issues a CA, TLS key and certificate. It stores them as entries in a Kubernetes Secret.
2) The KIM Snatch webhook mounts the Secret to make the CA, TLS key and certificate accessible through the local filesystem.
3) An webserver server ist started which exposes a HTTPS endpoint, using the mounted TLS certificate for securing incoming connections.
4) Changes of certificates are detected by a certificate watching library (Cert-Watcher).
5) Cert Watcher triggers a reload the HTTPS server configuration. This keeps the used TLS certificates up-to-date even after they were rotated.
6) Also the CA, used for creating the TLS certificates, is updated in the `WebhookConfiguration` by the Cert Watcher .
7) The `WebhookConfiguration` informs the Kubernetes API server about the existence of the webhook and refers to the CA. The API server uses CA to verify HTTPS connections established to the webhook.

> 8) API Server calls this HTTPS endpoint of the webhook for each received HTTP request. The webhook is allowed to modify the request before it's finally processed by the API server.

Reviewer: ```suggestion
8) The API Server calls this HTTPS endpoint of the webhook for each received HTTP request. The webhook is allowed to modify the request before the API server finally processes it.
```

|[Issuer CR](https://github.com/gardener/cert-management?tab=readme-ov-file#setting-up-issuers)|Custom resource (CR) used by Gardener cert-manager for managing certificates and CA.|
|[Certificate CR](https://github.com/gardener/cert-management?tab=readme-ov-file#requesting-a-certificate)|CR used by Gardener cert-manager for issuing certificates.|
|Kubernetes Secret|Created by Gardener cert-manager to store the generated CA, TLS key and certificate.|
|WebhookConfiguration|Kubernetes CR considered by the API server to invoke webhooks during request processing.|
|API Server|The Kubernetes API server that processes HTTP requests and applies resource changes in ETCD.|
|Webhook|Webserver intercepting requests before the API server processes them.|
|Webhook HTTPS-Server|Invoked by the API Server; intercepts all incoming requests of the API server.|
|Webhook certs volume|Mount point that makes the data entries of the Secret accessible over the webhook's filesystem.|
|Webhook Cert Watcher|Monitoring changes applied on the mounted Secret.|

## How It Works

> The activity diagram describes the internal technical process of Snatch more fine-grained:

Reviewer: ```suggestion
The activity diagram describes the internal technical process of KIM Snatch in more detail.
```
